doc/barclamp.md

```diff
@@ -0,0 +1,1177 @@
+  components to this node while all other OpenStack services reside on the main
+  controller. This greatly shortens the time a chef run for your barclamp's
+  controller side components takes and thus speeds up development. Make a
+  backup of this node as well to guard against breakage and allow for
+  clean-slate deployment of your barclamp's controller side components.
+
+## Development Environment
+
+Most of your development will take place on your Crowbar node (log in as root).
+You will find the currently running Crowbar code (both Rails application
+components and chef recipes) in the `/opt/dell` directory. We will provide
+detailed locations for all important components of a Barclamp in the _Basic
+Implementation_ section below.
+
+In order to track changes you make we recommend you keep a checkout of your
+`crowbar-openstack` fork on the Crowbar node and apply the changes you make in
+there
+
+### Preparations for Crowbar Node
+
+These are recommendations, but we'll assume you followed them for the remainder
+of this document since this shortens the commands in our examples. First of
+all, clone your `crowbar-openstack` fork to `/root/crowbar-openstack` on your
+Crowbar node and switch to your development branch.
+
+Next, create the following /root/.bashrc with a few useful aliases and
+environment variables (substitute with appropriate values where so indicated by
+comments):
+
+```
+# Replace "Crowbar Committer" by your name as it should appear in commit
+# messages.
+C_NAME="Crowbar Committer"
+
+# Replace "email@address.invalid" by your email address as it should appear in
+# commit messages.
+C_EMAIL="email@address.invalid"
+
+# Replace "6cefbd80254141a63f4c86df195cd930ac7eb10b" by your first commit's ID.
+export COMMIT="6cefbd80254141a63f4c86df195cd930ac7eb10b"
+
+# Insert a space delimited list of the Barclamps you are working on here.
+export BARCLAMPS="nova magnum"
+
+export GIT_AUTHOR_NAME="$C_NAME"
+export GIT_AUTHOR_EMAIL="$C_EMAIL"
+export GIT_COMMITTER_NAME="$C_NAME"
+export GIT_COMMITTER_EMAIL="#C_EMAIL"
+
+alias Cpatch="(cd ~/crowbar-openstack; git show | patch -p1 --merge -d /opt/dell)"
+
+# Full patch: applies your fork completely against /opt/crowbar
+function Fpatch()
+  {
+  cd ~/crowbar-openstack
+  git diff "$COMMIT^" HEAD | patch --merge -p1 -d /opt/dell
+  for f in /opt/dell/*.yml
+    do
+    ln -s $f /opt/dell/crowbar_framework/barclamps 2>&1 > /dev/null
+    done
+  chmod 755 /opt/dell/bin/*
+  }
+
+sync_crowbar()
+  {
+  barclamp_install.rb /opt/dell/crowbar_framework/barclamps/ &&
+  for barclamp in $BARCLAMPS
+    do
+    knife cookbook upload -o /opt/dell/chef/cookbooks/ mybarclamp
+    done
+  systemctl restart crowbar
+  }
+```
+
+We will make heavy use of the `Cpatch`, `Fpatch` and `sync_crowbar` commands
+thus defined in the following sections. A short explanation of the three:
+
+* `Fpatch` will apply your whole fork against `/opt/dell` using `patch(1)`. You
```

Comment:
Hmm. Now that I think about it again, it's got the same problem as the copy-stuff-to-`/opt/dell` approach: if one doesn't rebase religiously and frequently, old code in _other_ barclamps that is no longer around in `master` might creep in, leading to all sorts of weird problems that have nothing to do with the barclamp being developed.

For completeness' sake I mentioned the `barclamp_install` approach alongside the `cp` one now.